File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.js
const mergeKLists = (lists) => {
  if (!lists.length) return null;

  // sort
  lists.sort((a, b) => a && b ? a.val - b.val : a ? 1 : -1);

  const head = new ListNode();
  let tail = head;

  const append = (node) => {
    tail.next = node;
    tail = node;
    const next = tail.next;
    tail.next = null;
    return next;
  }

  const sort = (pointer) => {
    const list = lists.splice(pointer, 1)[0];
    let a = 0, b = lists.length - 1, c;
    while (a <= b) {
      c = ((a + b) / 2) | 0;
      lists[c] === null || lists[c].val <= list.val ? a = ++c : b = --c;
    }
    lists.splice(a, 0, list);
  }

  // build
  let pointer = 0;
  while (pointer < lists.length) {
    if (!lists[pointer]) {
      ++pointer;
      continue;
    }

    lists[pointer] = append(lists[pointer]);

    if (
      lists[pointer] &&
      lists[pointer + 1] &&
      lists[pointer].val > lists[pointer + 1].val
    ) {
      // resort
      sort(pointer);
    }
  }

  return head.next;
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.go
import (
	"slices"
	"container/heap"
)

type ListNodeHeap []*ListNode

func (h ListNodeHeap) Less(i, j int) bool {
	return h[i].Val < h[j].Val
}

func (h ListNodeHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h ListNodeHeap) Len() int {
	return len(h)
}

func (h *ListNodeHeap) Pop() (v any) {
	*h, v = (*h)[:h.Len()-1], (*h)[h.Len()-1]
	return
}

func (h *ListNodeHeap) Push(v any) {
	*h = append(*h, v.(*ListNode))
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
	lists = slices.DeleteFunc(lists, func(el *ListNode) bool { return el == nil })
	var h = new(ListNodeHeap)
	*h = ListNodeHeap(lists)
	heap.Init(h)
	head := &ListNode{}
	tail := head
	for h.Len() > 0 {
		node := heap.Pop(h).(*ListNode)
		if node.Next != nil {
			heap.Push(h, node.Next)
		}
		tail.Next = node
		tail = node
	}
	return head.Next
}